variant(wb55): add STeaMi board

nedseb opened the pull request:
Add support to the STeaMi board developed for the project I-NOVMICRO with help from ST Rousset.

I tried to follow all the steps of the Wiki. Let me know if something is missing.

fpistm commented:
Hi @nedseb 
Thanks for the PR.
Have you any documentations (User manual, schematics,...) helping to review the PR?

nedseb commented:
> Hi @nedseb Thanks for the PR. Have you any documentations (User manual, schematics,...) helping to review the PR?
Hi @fpistm,
You can find the reference design of the boards here : https://github.com/steamicc/steami-reference-design
I hope it will help you to review this PR. I will extract the pinout into a more human-readable version soon.

fpistm commented on variants/STM32WBxx/WB55R(C-E-G)V/variant_STEAM32_WB55RG.h:
```diff
@@ -0,0 +1,235 @@
+#endif
+
+/* backward compatibility */
+#define ARDUINO_PNUCLEO_WB55RG
```
Used for?

nedseb replied:
Come from [here](https://github.com/stm32duino/Arduino_Core_STM32/blob/ea74156acd823b6d14739f389e6cdc648f8ee36e/variants/STM32WBxx/WB55R(C-E-G)V/variant_P_NUCLEO_WB55RG.h#L135). I will remove it

nedseb commented:
I rebased this PR on top of the merge commit of PR #2528.

To be compatible with DAPLink, I need to replace `transport select hla_swd` with `transport select swd` (or at least remove it, because the default transport option on DAPLink is SWD).

To achieve this, I changed the default configuration of OpenOCD by modifying the value of `debug.server.openocd.scripts.1`. Previously, I used the target script to do this, but with the refactoring of the OpenOCD command, I broke (again) the CMake generation step of the CI.

I see 2 options, the first one is to modify the script `update_boarddb.py` to handle the configuration `STeaMi.menu.pnum.STEAM32_WB55RG.debug.server.openocd.scripts.1={runtime.platform.path}/debugger/select_swd.cfg` or simply remove [this line](https://github.com/stm32duino/Arduino_Core_STM32/blob/ea74156acd823b6d14739f389e6cdc648f8ee36e/platform.txt#L261) as the stlink interface configuration seems to don't need it anymore.

fpistm reviewed (commented):
First round.

nedseb commented on variants/STM32WBxx/WB55R(C-E-G)V/variant_STEAM32_WB55RG.h:
```diff
@@ -0,0 +1,235 @@
+#define P16 PE4
+#define P19 PC1
+#define P20 PC0
+#define P24 GPIO3
+#define P25 GPIO4
+#define P28 GPIO1
+#define P29 GPIO2
```
Not used on PinName but helps with the pin mapping verification

nedseb commented on variants/STM32WBxx/WB55R(C-E-G)V/variant_STEAM32_WB55RG.h:
```diff
@@ -0,0 +1,235 @@
+// On-board user button
+#define A_BUTTON     PA7
+#define B_BUTTON     PA8
+#define DOWN_BUTTON  GPIO6
+#define LEFT_BUTTON  GPIO7
```
Not used on PinName but helps with the pin mapping verification

nedseb commented on variants/STM32WBxx/WB55R(C-E-G)V/variant_STEAM32_WB55RG.h:
```diff
@@ -0,0 +1,235 @@
+#define DOWN_BUTTON  GPIO6
+#define LEFT_BUTTON  GPIO7
+#define MENU_BUTTON  PA0
+#define RIGHT_BUTTON GPIO5
+#define UP_BUTTON    GPIO8
```
Not used on PinName but helps with the pin mapping verification

fpistm commented on boards.txt:
```diff
@@ -13630,6 +13630,47 @@ ELV_Modular_System.menu.upload_method.serialMethod.upload.protocol=serial
+STeaMi.menu.pnum.STEAM32_WB55RG.build.series=STM32WBxx
+STeaMi.menu.pnum.STEAM32_WB55RG.build.product_line=STM32WB55xx
+STeaMi.menu.pnum.STEAM32_WB55RG.build.variant=STM32WBxx/WB55R(C-E-G)V
+STeaMi.menu.pnum.STEAM32_WB55RG.debug.server.openocd.scripts.0=interface/cmsis-dap.cfg
+STeaMi.menu.pnum.STEAM32_WB55RG.debug.server.openocd.scripts.1={runtime.platform.path}/debugger/select_swd.cfg
```
Did you test the debug using this config?

nedseb replied:
I tested it on the first version. I redoo a test in 5 min

fpistm replied:
OK Fine will test when get the board.

nedseb replied:
It works !
![image](https://github.com/user-attachments/assets/78fbd39a-1da2-4d49-b282-4cafee248306)

fpistm commented on boards.txt:
```diff
@@ -13630,6 +13630,47 @@ ELV_Modular_System.menu.upload_method.serialMethod.upload.protocol=serial
+STeaMi.name=STeaMi Board
+STeaMi.build.core=arduino
+STeaMi.build.variant_h=variant_{build.board}.h
+STeaMi.build.st_extra_flags=-D{build.product_line} {build.xSerial}
```
If I well understand there is no USB device support ?

nedseb replied:
Yes exactly. Like a microbit, you can access to the main MCU only through the onboard debug probe (called Interface MCU on microbit).

fpistm commented on variants/STM32WBxx/WB55R(C-E-G)V/variant_STEAM32_WB55RG.h:
```diff
@@ -0,0 +1,233 @@
+
+// GPIO expander
+// Upper pad
+#define GPIO1                   44 // P28
+#define GPIO2                   45 // P29
+#define GPIO3                   46 // P24
+#define GPIO4                   47 // P25
+
+// 4 directions switches
+#define GPIO5                   48
+#define GPIO6                   49
+#define GPIO7                   50
+#define GPIO8                   51
```
How do you plan to use those GPIO expander aliases?

fpistm replied:
outdated as no more value

nedseb replied:
I removed this definition. 

At the begining, I want to investigate a solution to hide the use of an IO Expander with for example a macro for `digitalRead` and `digitalWrite` but it's probably confusing on several use case of the Pin outside basic LED blinking. I'll make samples to show how to use the IO expander, it will be more interesting.

nedseb commented on variants/STM32WBxx/WB55R(C-E-G)V/variant_STEAM32_WB55RG.h:
```diff
@@ -0,0 +1,235 @@
+#define GPIO5 48
+#define GPIO6 49
+#define GPIO7 50
+#define GPIO8 51
```
I added a comment to state explicitly that pins GPIOX are only here for documentation purposes.

fpistm replied:
I saw and report it to the rebase. Anyway, I think it is quiet confusing and simply add those values as a comment:
```C
// GPIO expanders
// 4 directions switches
// GPIO5
// GPIO6
// GPIO7
// GPIO8
```

 but up to you :wink:

nedseb replied:
Ok for me !

fpistm reviewed (commented):
I've squashed and rebased on top of the main with some fix and formatting.
Please fetch your repo befopre any further change.

About the GPIO expander definition, I don't think it is a good idea to give them a number.


Finally, for pins not on a connector, you could add a custom peripheral pins to save space.

fpistm commented:
I've made a commit to fix the cmake issue.

fpistm reviewed (approved):
LGTM
